### PR TITLE
fix(train): enable BF16 on validated MPS trainers

### LINKED ISSUE #563: Resident Transformers trainers ignore BF16-capable MPS and run in FP32
## Reproduction

I reproduced this on Apple Silicon using current `main` (`6c13c44`), PyTorch 2.12.1 and Transformers 5.12.1. Four one-step hardware smokes all completed on `mps`, but Soup disabled mixed precision and kept the resident training path in FP32:

| Task | `TrainingArguments` | Model parameters | Trainable parameters | Result |
| --- | --- | --- | --- | --- |
| SFT LoRA | `bf16=false`, `fp16=false` | FP32 | FP32 | 1 step, loss 10.8306 |
| DPO LoRA | `bf16=false`, `fp16=false` | FP32 | FP32 | 1 step, loss 0.6931 |
| Reward model LoRA | `bf16=false`, `fp16=false` | FP32 | FP32 | 1 step, loss 0.6865 |
| PRM | `bf16=false`, `fp16=false` | FP32 | FP32 | 1 step, loss 1.3869 |

The same runtime successfully executes native MPS BF16 matmul, backward, synchronization and `torch.autocast(device_type="mps", dtype=torch.bfloat16)`. `TrainingArguments(bf16=True)` also resolves to device `mps` without error.

This is distinct from #516/#519. That fix preserves BF16 in the layer-streaming store. The resident Transformers trainer path still routes every non-CUDA device through `bf16_fp16_flags -> (False, False)`. PRM additionally hardcodes a FP32 load for every non-CUDA device and never passes precision flags to `TrainingArguments`.

## Impact

On a large BF16 checkpoint such as Qwen3.8-27B, the Mac path can lose the memory and compute benefits expected from BF16. It also makes the Apple Silicon behavior disagree with the checkpoint dtype and with the already-corrected streaming path.

## Expected behavior

- A live MPS runtime that can allocate and execute BF16 should select `bf16=true`, `fp16=false` for Transformers trainers.
- CPU and an unavailable or BF16-incapable MPS runtime must remain FP32.
- CUDA capability routing must remain unchanged.
- SFT, DPO, reward-model and PRM one-step hardware smokes must complete with the resolved BF16 policy.
- The MPS capability decision should be shared with layer streaming rather than duplicated.
- Platform-independent tests must pin both supported and unsupported MPS paths; the real hardware smoke should record the runtime and dtype evidence.

## Ownership

I am taking this issue and will prepare a focused fix with before/after Apple Silicon evidence.

### PR DESCRIPTION
**Pull request by Amix29**
## Summary

- add a live MPS BF16 capability probe shared by resident training and layer streaming
- opt only the hardware-validated text trainers into MPS BF16: SFT, DPO, reward modelling, and PRM
- keep CPU and unvalidated MPS trainers on their existing FP32 path
- keep PRM master weights in FP32 while using BF16 autocast, avoiding a measured fatal Metal optimizer dtype mismatch
- document the MPS/Transformers precision policy and its boundary

Closes #563.

## Apple Silicon evidence

Validated on an Apple M4 Max with PyTorch 2.12.1 and Transformers 5.12.1.

Before this change, one-step SFT, DPO, reward-model, and PRM runs all completed on MPS with `bf16=false` and `fp16=false`.

After this change:

| Task | Precision policy | Storage evidence | One-step result |
| --- | --- | --- | --- |
| SFT LoRA | `bf16=true`, `fp16=false` | 134,515,008 frozen base parameters in BF16; 230,400 trainable LoRA parameters in FP32 | completed, loss 0.873974 |
| DPO LoRA | `bf16=true`, `fp16=false` | 134,515,008 frozen base parameters in BF16; 230,400 trainable LoRA parameters in FP32 | completed, loss 0.677233 |
| Reward model LoRA | `bf16=true`, `fp16=false` | 134,515,584 base parameters in BF16; 230,976 trainable parameters in FP32 | completed, loss 0.755024 |
| PRM | `bf16=true`, `fp16=false` | FP32 master parameters with BF16 autocast | completed, loss 1.386875 |

The first PRM attempt also loaded its trainable base in BF16. Metal aborted in matrix multiplication because the destination and accumulator dtypes differed. The submitted policy therefore preserves FP32 PRM master weights and enables only the hardware-validated BF16 autocast path. This PR makes no throughput claim from these single-step smokes.

## Validation

- `pytest -q tests/test_issue563_mps_bf16_training.py tests/test_issue385_stream_dtype.py tests/test_v0533.py tests/test_v05311.py --no-cov`: 152 passed, 4 skipped
- `pytest tests/ -q --tb=short --no-cov`: 18,942 passed, 82 skipped, 5 deselected
- `ruff check src/soup_cli/ tests/`
- `git diff --check`

### CODE REVIEW

**Comment by MakazhanAlpamys** [comment]
Thank you, @Amix29. #564 was squash-merged. BF16 is enabled through a runtime probe rather than a platform string, and the necessary PRM FP32 master-weight exception remains. I ran 155 focused tests; forcing the MPS BF16 decision false failed the named capability test. Docs, changelog, and all 13 CI jobs are green.
